_includes/v20.2/sql/role-considerations.md

```diff
@@ -0,0 +1,12 @@
+- Role names (also known as usernames):
+    - Are case-insensitive.
+    - Can contain only letters, numbers, underscores (`_`), and periods. When a period is present, single or double quotes around the name are required in SQL commands.
```

Comment:
@knz How do you define "Valid SQL identifier"? Valid per the SQL Standard?

Reply:
per the postgres lexical rules

Reply:
@knz OK thanks.

## Naming details edit

I've reworked this in light of our comments, further testing, and to match the sensible [Postgres docs](https://www.postgresql.org/docs/current/sql-syntax-lexical.html) paradigm of splitting into what's allowed _without quotes_ vs. _with quotes_. I include 'with quotes' first, because we recommend always using quotes, and this makes it easier to see the short list of character usage that's lost without quotes. How's this?

(Note: I'm temporarily removing the 63-character limit mention; I have a follow-up question about that.)

> - Role names (also known as usernames) are case-insensitive and must be unique.
> - When surrounded by quotes in SQL statements (always recommended)—single or double quotes, depending on the statement—role names:
>     - Can contain letters, underscores, digits, periods, and dashes. Letters may include `a`-`z`, those with diacritical marks, and non-Latin letters*.
>     - Can begin with a letter, underscore, or digit.
> - When referenced without quotes, role names:
>     - Cannot contain periods or dashes.
>     - Cannot begin with a digit.
>     - Cannot match the name of a SQL keyword.

\* I'm provisionally adding this for your review, in case it's accurate, because it is specified in the Postgres docs, though I'm actually not sure which characters it's referring to.

## Length - characters vs bytes

Upon further testing, I think we are limiting name length by bytes and not characters, and that some characters are using two bytes each.

Testing a username that comprises 35 instances of `å` (with or without quotes) gets me this:

```
ERROR: username "ååååååååååååååååååååååååååååååååååå" is too long
```

I am wondering if this is a bug, and we should be able to store `å` using a single byte, or if it's expected. And regardless, whether we should add a caveat to the docs.

Side note: the hint in the SQL shell displayed when the name is too long does match your overall summary of requirements of characters that are allowed with quotes: "Usernames are case insensitive, must start with a letter, digit or underscore, may contain letters, digits, dashes, periods, or underscores, and must not exceed 63 characters." 

* However
   * the "63 characters" point is now in question. 
   * that this can be displayed when not using quotes (e.g. via `CREATE USER ååååååååååååååååååååååååååååååååååå`') and yet refers to character usage that won't work without quotes, may be confusing for users.

## Note on unicode

Just FYI, we appear to allow unicode characters via statements like `CREATE ROLE "\U+AB3E;"`.

That format isn't documented in Postgres, and I'm not sure how you'd string together multiple characters. We do not appear to support the similar Postgres format [described here](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#identifiers-and-key-words#:~:text=A%20variant%20of%20quoted%20identifiers%20allows%20including,Unicode%20characters%20identified%20by%20their%20code%20points.).

Reply:
> How's this?

probably fine. would like to see it in context though.

> I am wondering if this is a bug

Short answer: the UX is inconsistent and that's the bug. We can either choose 63 bytes or 63 characters (both would be possible in the code), but we need to ensure that the implementation matches the description/doc.

Which one *should* it be? I don't know. I believe we need a limit of 63 bytes for pure ASCII names, but if the name is pure ASCII then 63 bytes is also == 63 characters. Maybe start by documenting the 63 byte limit, and we can discuss later about extending the limit in the code to 63 characters.

> we appear to allow unicode characters via statements like CREATE ROLE "\U+AB3E;".

that's a happy accident, however probably not one we want to document yet, in light of this:

> That format isn't documented in Postgres, and I'm not sure how you'd string together multiple characters. We do not appear to support the similar Postgres format described here.

I would say we should aim to do what postgres accepts here. If we're not there yet, we need to file a followup issue in the crdb repo, and document the mismatch as a known limitation.

Reply:
Documenting the current state, plus any differences with Postgres, sounds good to me. Created an issue https://github.com/cockroachdb/docs/issues/10010 and included one other difference that I noticed:

> Difference in how quoted names are treated: always case insensitive in CRDB vs. case sensitive _when created using quotes_ in PostgreSQL.